### PR TITLE
feat(graphstore): introduce GraphStore abstraction and add InMemory implementation

### PR DESCRIPTION
**Pull request by sanket-rajput**
## What does this PR do?

This PR introduces a new `GraphStore` abstraction to decouple knowledge
graph storage from its concrete implementation.

The goal is to make the knowledge graph pluggable, extensible, and
future-proof for alternative backends (e.g., persistent or distributed stores).

---

### Introduced `GraphStore` trait with:
- CRUD operations (`upsertNode`, `upsertEdge`, `deleteNode`, `deleteEdge`)
- Traversal API (direction-aware, maxDepth-controlled)
- Query API via `GraphFilter`
- Graph statistics via `GraphStats`

### Added supporting models:
- `Direction`
- `GraphFilter`
- `TraversalConfig`
- `GraphStats`

### Implemented `InMemoryGraphStore`:
- Full `GraphStore` API support
- BFS-based traversal with depth control
- Query filtering support
- Proper edge cleanup on node deletion

### Updated `JsonGraphStore`:
- Aligned implementation with new `GraphStore` interface

### Tests:
- Added unit tests for `InMemoryGraphStore`

All existing tests pass.

---

## Related issue

Relates to #217 (Knowledge Graph Module)

---

## How was this tested?

Ran:

- `sbt scalafmtAll`
- `sbt +test`

Results:
- 3377 tests succeeded
- 0 failed

New tests cover:
- Node insertion & retrieval
- Edge validation
- Node deletion cleanup
- Graph statistics
- Traversal with `maxDepth`

### CODE REVIEW

**Review by rorygraves — COMMENTED** [review]
## Review: APPROVE WITH MINOR CHANGES

Well-structured PR introducing the `GraphStore` trait and `InMemoryGraphStore` implementation for the Knowledge Graph epic (#650). The trait API matches the design spec, `Result[A]` is used correctly throughout, and CI passes on both Scala versions and all platforms. Good work!

### Must-fix (2 items)

#### 1. Thread safety on `InMemoryGraphStore`
`InMemoryGraphStore` uses an unsynchronized `private var graph` which is a data race if accessed from multiple threads. At minimum, add a scaladoc note that this implementation is **not thread-safe**. Ideally, use a synchronized wrapper or atomic reference if concurrent access is expected.

#### 2. Missing test coverage for `JsonGraphStore`
The PR adds 10 new methods to `JsonGraphStore` but zero tests for them. Codecov reports only 35.8% patch coverage on that file with 52 uncovered lines. Please add tests covering at minimum: `addNode`, `addEdge`, `getNeighbors`, `traverse`, and `query` for the JSON variant.

### Should-fix (3 items)

3. **Traversal algorithm mismatch** — `InMemoryGraphStore.traverse()` uses DFS while `JsonGraphStore.traverse()` uses BFS. Both implement the same trait method but return results in different orders. Pick one algorithm (BFS is more standard for graph traversal) and use it consistently.

4. **Inconsistent error on missing start node** — `InMemoryGraphStore.traverse()` returns `Left(ProcessingError)` for a missing start node, but `JsonGraphStore.traverse()` returns `Right(Seq.empty)`. These should behave the same.

5. **`JsonGraphStore.query` ignores property filters** — `propertyKey`/`propertyValue` parameters are silently ignored in the JSON implementation while `InMemoryGraphStore` supports them. Either implement property filtering or return an error indicating it's unsupported.

### Nits

6. Fragile property value comparison using `.toString.replace("\"", "")` — consider pattern matching on the `ujson.Value` type instead.
7. No ScalaDoc on the new public API types (`GraphStore`, `GraphNode`, `GraphEdge`, etc.).
8. Test class uses `Test` suffix (`InMemoryGraphStoreTest`) — project convention appears to use `Spec` suffix.

Overall the architecture is solid. Once the must-fix items are addressed, this should be ready to merge.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/InMemoryGraphStore.scala [hunk 1]
```diff
@@ -0,0 +1,134 @@
+package org.llm4s.knowledgegraph.storage
+
+import org.llm4s.knowledgegraph.{ Graph, Node, Edge }
+import org.llm4s.types.Result
+import org.llm4s.error.ProcessingError
+
+class InMemoryGraphStore extends GraphStore {
+
+  private var graph: Graph = Graph.empty
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/InMemoryGraphStore.scala]: This still uses unsynchronized mutable state (`private var graph`) and remains racy under concurrent access. Please either guard updates (synchronization/atomic) or clearly document this class as not thread-safe. Principle: mutable shared state must be explicitly concurrency-safe.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/JsonGraphStore.scala [hunk 2]
```diff
@@ -78,18 +79,128 @@ class JsonGraphStore(path: Path) extends GraphStore {
...
+
+      Graph(filteredNodes, filteredEdges)
+    }
+
+  override def traverse(
+    startId: String,
+    config: TraversalConfig
+  ): Result[Seq[Node]] =
+    loadFromFile().map { g =>
+      if (!g.hasNode(startId)) Seq.empty
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/JsonGraphStore.scala]: Missing-start behavior is inconsistent with `InMemoryGraphStore`: here it returns empty result, there it returns `Left(ProcessingError)`. Please align both to one contract. Principle: error model must be predictable across implementations.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/InMemoryGraphStore.scala [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  }
+
+  /* ------------------- Traversal ------------------- */
+
+  override def traverse(
+    startId: String,
+    config: TraversalConfig
+  ): Result[Seq[Node]] = {
+
+    def dfs(current: String, depth: Int, visited: Set[String]): Set[String] =
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/InMemoryGraphStore.scala]: Traversal strategy is DFS here, while `JsonGraphStore.traverse` is BFS-style frontier expansion. Same trait API should not return order/coverage inconsistently across implementations. Principle: contract-consistent behavior across implementations.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/JsonGraphStore.scala [hunk 4]
```diff
@@ -78,18 +79,128 @@ class JsonGraphStore(path: Path) extends GraphStore {
...
+          case Direction.Both if e.source == nodeId =>
+            g.nodes.get(e.target).map(n => (e, n))
+          case Direction.Both if e.target == nodeId =>
+            g.nodes.get(e.source).map(n => (e, n))
+          case _ => None
+        }
+      }
+    }
+
+  override def query(filter: GraphFilter): Result[Graph] =
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/JsonGraphStore.scala:133]: `propertyKey` / `propertyValue` from `GraphFilter` are ignored here, but `InMemoryGraphStore.query` applies them. Please implement equivalent filtering or return a clear unsupported error. Principle: query semantics should be uniform across backends.

File: modules/core/src/test/scala/org/llm4s/knowledgegraph/storage/InMemoryGraphStoreTest.scala [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    val n1    = Node("1", "Person")
+    val n2    = Node("2", "Person")
+    val edge  = Edge("1", "2", "KNOWS")
+
+    store.upsertNode(n1)
+    store.upsertNode(n2)
+
+    store.upsertEdge(edge).isRight shouldBe true
+
+    val neighbors = store.getNeighbors("1", Direction.Outgoing).toOption.get
```
Comment (vim89) [inline modules/core/src/test/scala/org/llm4s/knowledgegraph/storage/InMemoryGraphStoreTest.scala]: Please avoid unsafe `.toOption.get` in tests; use `fold`/pattern-based assertions to keep tests total and refactor-safe. Principle: avoid partial functions even in test code.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/InMemoryGraphStore.scala [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  override def query(filter: GraphFilter): Result[Graph] = {
+    val filteredNodes = graph.nodes.values
+      .filter { node =>
+        val labelMatch =
+          filter.label.forall(_ == node.label)
+
+        val propertyMatch =
+          (filter.propertyKey, filter.propertyValue) match {
+            case (Some(k), Some(v)) =>
+              node.properties.get(k).exists(_.toString.replace("\"", "") == v)
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/InMemoryGraphStore.scala]: This string-based property comparison is brittle (`toString.replace`) and type-unsafe. Please match on `ujson.Value` shape directly for deterministic property filtering. Principle: typed value handling over stringly comparisons.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/InMemoryGraphStore.scala (hunk 1 above)
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/InMemoryGraphStore.scala]: `private var graph` is still unsynchronized shared mutable state.
Please make concurrency behavior explicit (document non-thread-safe) or guard mutations.
Principle: shared mutable state must be concurrency-safe by contract.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/InMemoryGraphStore.scala (hunk 3 above)
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/InMemoryGraphStore.scala]: Traversal here is DFS recursion, while `JsonGraphStore` uses frontier/BFS style.
Same `GraphStore` API should not change traversal semantics by backend.
Principle: implementation parity is part of API correctness.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/JsonGraphStore.scala (hunk 4 above)
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/JsonGraphStore.scala:133]: `GraphFilter.propertyKey/propertyValue` are still ignored in this implementation.
Either implement equivalent filtering or return a clear unsupported error.
Principle: query semantics must be uniform across storage backends.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/JsonGraphStore.scala (hunk 2 above)
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/JsonGraphStore.scala]: Missing-start behavior diverges: this returns empty result, while in-memory returns `Left(ProcessingError)`.
Please align both implementations to one contract.
Principle: error model must be deterministic across implementations.

File: modules/core/src/test/scala/org/llm4s/knowledgegraph/storage/InMemoryGraphStoreTest.scala (hunk 5 above)
Comment (vim89) [inline modules/core/src/test/scala/org/llm4s/knowledgegraph/storage/InMemoryGraphStoreTest.scala]: This uses partial extraction via `.toOption.get`.
Please assert with `fold` or explicit pattern to keep tests total and refactor-safe.
Principle: avoid partial operations even in test code.

**Comment by sanket-rajput** [comment]
Hi @rorygraves , @vim89 I’ve pushed an update addressing the requested changes. 
Would appreciate a re-review when you have time 🙌

Updates included:

• Thread safety : InMemoryGraphStore now uses AtomicReference with CAS loops for linearizable updates (removed shared mutable var usage).
• Traversal parity : standardized traversal behavior (BFS-style frontier expansion) to align with JsonGraphStore.
• Missing-start behavior : both implementations now return Left(ProcessingError) for consistency.
• Property filtering : JsonGraphStore now supports propertyKey/propertyValue filtering.
• Tests : added JsonGraphStoreTest to improve coverage and validate new API paths.

Please let me know if anything still looks inconsistent with the GraphStore contract.
Thanks again for the detailed feedback!

**Review by rorygraves — CHANGES_REQUESTED** [review]
Thank you for the work on this PR — the lock-free CAS `InMemoryGraphStore` is genuinely the highest-quality piece of code across both competing GraphStore PRs, and the functional BFS in `JsonGraphStore.traverse` is clean. However, after reviewing both #666 and #710 together, we've decided to take #710 forward as the base implementation because it includes the `GraphStoreConsistencySpec` shared test matrix and the `KnowledgeGraphGenerator` wiring that we need.

**We're asking you to close this PR**, but we want to acknowledge your contributions clearly:

- **Your CAS `InMemoryGraphStore` pattern is being adopted into #710** — specifically the `upsertEdge` node-existence re-check inside each CAS retry loop, which handles the concurrent `deleteNode` race correctly. We've left a comment on #710 asking the author to adopt it.
- **Your functional BFS design** informed our review of #710's traversal implementation.
- **Your test coverage of `JsonGraphStore`** identified gaps that #710 also needs to address.

If you'd like to contribute further, the `SQLiteGraphStore` backend that #710 is being asked to remove would be a great standalone follow-up PR in a separate optional module — and your understanding of the `GraphStore` trait would give you a head start on that.

Thank you again for the quality of the implementation.

**Comment by rorygraves** [comment]
Closing in favour of #710 as discussed in the review above. Your CAS implementation pattern and BFS design are being carried forward — thank you for the contribution.

**Comment by sanket-rajput** [comment]
Thanks for the detailed review and feedback, really appreciate it.

Glad to hear the CAS update pattern and traversal design were useful and are being carried forward 👍

I’ll follow #710 and look for opportunities to contribute further, the optional SQLiteGraphStore module sounds interesting, so I may explore that as a follow-up PR.

Thanks again for the guidance!
